test/type-error.js

```diff
@@ -0,0 +1,563 @@
+				return new Error('This is fo FooBar', foo);
+			}
+			return foo;
+		}`,
+		`if (Array.isArray(foo) || ArrayBuffer.isView(foo)) {
+			lastError = new Error('This is fo FooBar', foo);
+		}`,
+		`if (!isFinite(foo)) {
+			throw new TypeError();
+		}`,
+		`if (isNaN(foo)) {
+			throw new TypeError();
+		}`,
+		`if (foo instanceof boo) {
+			throw new TypeError();
+		}`,
+		`if (typeof boo === 'Boo') {
+			throw new TypeError();
+		}`,
+		`if (Number.isNaN(foo)) {
+			throw new TypeError();
+		}`,
+		`if (Number.isFinite(foo) && Number.isSafeInteger(foo) && Number.isInteger(foo)) {
+			throw new TypeError();
+		}`,
+		`if (Array.isArray(foo) || (Blob.isBlob(foo) || Blip.isBlip(foo))) {
+			throw new TypeError();
+		}`,
+		`if (typeof foo === 'object' || (Object.isFrozen(foo) || 'String' === typeof foo)) {
+			throw new TypeError();
+		}`,
+		`if (isNaN) {
+			throw new Error();
+		}`,
+		`if (isObjectLike) {
+			throw new Error();
+		}`,
+		`if (isNaN.foo()) {
+			throw new Error();
+		}`,
+		`if (typeof foo !== 'object' || foo.bar() === false) {
+			throw new TypeError('Expected Foo being bar!');
+		}`,
+		`if (foo instanceof Foo) {
+			throw new TypeError('Expected Foo being bar!');
+		}`,
+		`if (!foo instanceof Foo) {
+			throw new TypeError('Expected Foo being bar!');
+		}`,
+		`if (foo instanceof Foo === false) {
+			throw new TypeError('Expected Foo being bar!');
+		}`,
+		`throw new Error('💣')`,
+		`if (!Number.isNaN(foo) && foo === 10) {
+			throw new Error('foo is not 10!');
+		}`,
+		`function foo(foo) {
+			if (!Number.isNaN(foo) && foo === 10) {
+				timesFooWas10 += 1;
+				if (calculateAnswerToLife() !== 42) {
+					openIssue('Your program is buggy!');
+				} else {
+					return printAwesomeAnswer(42);
+				}
+				throw new Error('foo is 10');
+			}
+		}`,
+		`function foo(foo) {
+			if (!Number.isNaN(foo)) {
+				timesFooWas10 += 1;
+				if (calculateAnswerToLife({with: foo}) !== 42) {
+					openIssue('Your program is buggy!');
+				} else {
+					return printAwesomeAnswer(42);
+				}
+				throw new Error('foo is 10');
+			}
+		}`,
+		`if (!x.isFudge()) {
+			throw new Error('x is no fudge!');
+		}`,
+		`if (!_.isFudge(x)) {
+			throw new Error('x is no fudge!');
+		}`
+	],
+	invalid: [
+		{
+			code: `if (Array.isArray(foo)) {
+				throw new Error('foo is an Array');
+			}`,
+			output: `if (Array.isArray(foo)) {
+				throw new TypeError('foo is an Array');
+			}`,
+			errors
+		},
+		{
+			code: `if (foo instanceof bar) {
+				throw new Error(foobar);
+			}`,
+			output: `if (foo instanceof bar) {
+				throw new TypeError(foobar);
+			}`,
+			errors
+		},
+		{
+			code: `if (_.isElement(foo)) {
+				throw new Error();
+			}`,
+			output: `if (_.isElement(foo)) {
+				throw new TypeError();
+			}`,
+			errors
+		},
+		{
+			code: `if (_.isElement(foo)) {
+				throw new Error;
+			}`,
+			output: `if (_.isElement(foo)) {
+				throw new TypeError;
+			}`,
+			errors
+		},
+		{
+			code: `if (wrapper._.isElement(foo)) {
```

Comment:
The `wrapper` tests for property-chains ending in a type checking identifier. I added that one when realizing that 1) i can't be sure the hosting objects are named as we would expect and 2) type checking functions might be attached to a property chain.

But writing this reminds me of checking for calling-chains, too. 🤔 

I did think about checking for Object-names too, but i can't think of any reliable and performant solution being able to handle all cases of attachments.